### PR TITLE
Hide master bar on Jetpack Cloud Pricing page

### PR DESCRIPTION
**Pull request by manzoorwanijk**
Fixes `1164141197617539-as-1201622749500765`

#### Changes proposed in this Pull Request

* Hide master bar on Jetpack cloud pricing page using CSS instead of toggling its visibility via redux.
* This is to ensure that master bar is hidden on/before page load.

#### Testing instructions

* Checkout this branch and run `yarn start` before running `yarn start-jetpack-cloud-p`
* Goto http://jetpack.cloud.localhost:3001/pricing
* Throttle the connection in **Network** tab using **Slow 3G**
* Reload the page to ensure that you don't see the Calypso blue masterbar
* Goto http://jetpack.cloud.localhost:3001 again and select a site
* Goto Backup, Scan and Search sections
* Ensure that master bar at the top is not hidden.

### Testing instructions for devs
* Goto https://cloud.jetpack.com/pricing
* Open Dev Tools > Redux
* In filter input, enter `MASTERBAR_TOGGLE_VISIBILITY`
* Hover over to the action item below and click on **Jump**
* Click on the Forward and Backward buttons to toggle the visibility of the master bar
* Notice how the visibility of master bar at the top is toggled
* Now goto http://jetpack.cloud.localhost:3001/pricing
* Follow the same steps after opening Dev tools > Redux
* Notice how master bar is hidden regardless of its visibility state

Before
image

After
image

### CODE REVIEW

**Review by mattgawarecki — COMMENTED** [review]
First, from a "shipping is good" perspective: this definitely accomplishes our goal. I see we're definitely hiding the masterbar via CSS, which removes any chance of a user seeing it when they shouldn't. _But_, I wonder if telling React not to render it in the first place would be a cleaner, more performant approach.

For example, in [`client/jetpack-cloud/sections/pricing/controller.tsx`](https://github.com/Automattic/wp-calypso/blob/5fcd9d6c1f4e3d917c780eb903441414d7017228/client/jetpack-cloud/sections/pricing/controller.tsx#L33), could we make the assignment to `context.nav` conditional on the page URL not being part of the `/pricing` path? I recognize a simple `if` statement here could be fragile, but I see similar spots in the same file; maybe there's a way to separate the pricing-specific logic into its own function?

**Comment by manzoorwanijk** [comment]
> I wonder if telling React not to render it in the first place would be a cleaner, more performant approach.

I agree.

> For example, in [`client/jetpack-cloud/sections/pricing/controller.tsx`](https://github.com/Automattic/wp-calypso/blob/5fcd9d6c1f4e3d917c780eb903441414d7017228/client/jetpack-cloud/sections/pricing/controller.tsx#L33), could we make the assignment to `context.nav` conditional on the page URL not being part of the `/pricing` path?

Actually the whole calypso routing is still a mystery to me. 😄 

I tried to replace this line with
https://github.com/Automattic/wp-calypso/blob/4ec1da22f8dce740a23fdbdb371b3afa482a7bee/client/layout/index.jsx#L352
with
```js
const noMasterbarForSection = [ 'signup', 'jetpack-connect', 'jetpack-cloud-pricing' ].includes( sectionName );
```

But that section name is set pretty late and master bar is still shown until the section loads.

Even `currentRoute` there is not set before section loads.

File: client/jetpack-cloud/sections/pricing/header/style.scss [hunk 1]
```diff
@@ -1,6 +1,10 @@
 @import '@wordpress/base-styles/breakpoints';
 @import '@wordpress/base-styles/mixins';
 
+#header {
+	display: none;
+}
```
Comment (jsnajdr) [inline client/jetpack-cloud/sections/pricing/header/style.scss]: This CSS rule says that the `#header` element should _always_ be hidden. It's not scoped to any section. And as soon as the stylesheet is loaded, it will indeed hide it. And it will remain hidden even if we navigate away from the Pricing section to another route, even one that wants to show the masterbar.

Comment (jsnajdr) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: The Jetpack Cloud Pricing handler already disables the masterbar here:

https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/pricing/controller.tsx#L32

Do you know why it doesn't work?

**Review by jsnajdr — CHANGES_REQUESTED** [review]

File: client/jetpack-cloud/sections/pricing/header/style.scss (hunk 1 above)
Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: > And as soon as the stylesheet is loaded, it will indeed hide it. And it will remain hidden even if we navigate away from the Pricing section to another route, even one that wants to show the masterbar.

Oh, thanks. Didn't realize that but it does show up when you navigate away from pricing page. 🤔 🤷 

> Do you know why it doesn't work?

Please see the [comment](https://github.com/Automattic/wp-calypso/pull/59806#issuecomment-1007181077) above. That masterbar is disabled pretty late and on a slow connection, users may see that bar until the page loads.

Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: Pushed a fix to restrict the CSS rule to pricing section. Thanks

Comment (jsnajdr) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: This is better now, but I'd still like to understand better why other solutions don't work. First, I can't reproduce the issue even on a Slow 3G connection. The masterbar never renders on `/pricing` load. Maybe it does on a navigation from/to some other page? Do you have a video of the bug?

Also, the solution that adds `jetpack-cloud-pricing` to `noMasterbarForSection` should also work. The `.is-section-jetpack-cloud-pricing` body CSS class is added by the `Layout` component literally at the same time, during the same render, when the `jetpack-cloud-pricing` section is being activated. And it does that in a way that's established and used by other parts of Calypso.

Another suspicious thing I found that might be related is [this `showMasterbar` effect](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans/jetpack-plans/selector.tsx#L112) in the plans selector component, added by @monsieur-z in #59324. Is it possible that the `SelectorPage` component unmounts while the `/pricing` route is still active? An individual component somewhere down the tree shouldn't do `hideMasterbar` or `showMasterbar`. Only a route handler or the `Layout` component should do that because it's a responsibility of the top-level page layout.

To summarize, this patch is mergeable and does the job (or at least doesn't do harm) but I believe there should be a better solution and would like to understand the problem better.

Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: Steps for devs (See the GIF attached):

- Goto https://cloud.jetpack.com/pricing
- Open Dev Tools > Redux
- In filter input, enter `MASTERBAR_TOGGLE_VISIBILITY`
- Hover over to the action item below and click on **Jump**
- Click on the **Forward** and **Backward** buttons to toggle the visibility of the masterbar


DEMO
demo

Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: > To summarize, this patch is mergeable and does the job (or at least doesn't do harm) but I believe there should be a better solution and would like to understand the problem better.

Yeah, I agree. But I am new here and not a master of calypso routing yet. 😄

Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: > The `.is-section-jetpack-cloud-pricing` body CSS class is added by the `Layout` component literally at the same time, during the same render, when the `jetpack-cloud-pricing` section is being activated.

I see that `is-section-jetpack-cloud-pricing` class is added before anything is displayed on the page.

Comment (jsnajdr) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: > Steps for devs (See the GIF attached):

That's an artificial scenario though, that does time-travel in the Redux devtool. Can I reproduce the bug by simply loading the `/pricing` page and observing it?

Before this patch, there should appear an empty page with a Jetpack masterbar, and then the pricing page without the masterbar, is that right? There was never a "pricing page with masterbar" combination, right?

When I'm trying to reproduce this, I only ever see an "empty page", never an "empty page with masterbar". I know that such a page is rendered by React at least once, and Redux time-travel shows that, but it's never painted because it's immediately (within the same event loop tick) superseded by an updated render.

> not a master of calypso routing yet.

Don't worry about it, nobody is a master of that 🙂

Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: > Can I reproduce the bug by simply loading the `/pricing` page and observing it?

Yes, it was reported here p8oabR-LV-p2#comment-5760 and I can confirm it using an ultra slow custom throttle in my browser.

Comment (rcanepa) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: I haven't verified this yet, but could it be that users who have access to the Partners Portal are seeing that?

Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: > could it be that users who have access to the Partners Portal are seeing that?

No, it's shown regardless.

Comment (rcanepa) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: Ok. I'm seeing it now. One has to start from a site. In other words, there must be a site in the URL (i.e., the path must look like `/pricing/:site`).

Comment (rcanepa) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: I suspect that having a site in the URL makes the app dispatch some requests, which delay the entire rendering process, and one is able to see the masterbar for a brief moment.

Comment (jsnajdr) [reply client/jetpack-cloud/sections/pricing/header/style.scss]: Yeah I noticed that Eric's P2 report uses a `/pricing/:site` URL, not just `/pricing`.

**Review by jsnajdr — APPROVED** [review]
This is good as a bug fix :+1: I have some ideas how to improve Masterbar hiding/showing on the framework level, and how to migrate that functionality away from Redux, but that's quite beyond the scope of this particular PR.
